README.md
# Rust VEML6075 UVA and UVB Light Sensor Driver

[![crates.io](https://img.shields.io/crates/v/veml6075.svg)](https://crates.io/crates/veml6075)
[![Docs](https://docs.rs/veml6075/badge.svg)](https://docs.rs/veml6075)
[![Build Status](https://github.com/eldruin/veml6075-rs/workflows/Build/badge.svg)](https://github.com/eldruin/veml6075-rs/actions?query=workflow%3ABuild)
[![Coverage Status](https://coveralls.io/repos/github/eldruin/veml6075-rs/badge.svg?branch=master)](https://coveralls.io/github/eldruin/veml6075-rs?branch=master)

This is a platform agnostic Rust driver for the VEML6075 UVA and UVB light sensor,
based on the [`embedded-hal`](https://github.com/rust-embedded/embedded-hal) traits.

This driver allows you to:
- Enable/disable the sensor. See: `enable()`.
- Read calibrated UVA, UVB and UV index measurement. See: `read()`.
- Read raw measurement. See: `read_uva_raw()`.
- Set integration time. See: `set_integration_time()`.
- Set dynamic setting. See: `set_dynamic_setting()`.
- Change operating mode. See: `set_mode()`.
- Trigger measurement when on active force mode. See: `trigger_measurement()`.
- Read the device id. See: `read_device_id()`.

[Introductory blog post](https://blog.eldruin.com/veml6075-uva-uvb-uv-index-light-sensor-driver-in-rust/)

## The device
The VEML6075 senses UVA and UVB light and incorporates photodiode, amplifiers,
and analog / digital circuits into a single chip using a CMOS process. When the
UV sensor is applied, it is able to detect UVA and UVB intensity to provide a
measure of the signal strength as well as allowing for UVI measurement.

The VEML6075 provides excellent temperature compensation capability for keeping
the output stable under changing temperature. VEML6075's functionality is easily
operated via the simple command format of I2C (SMBus compatible) interface protocol.
VEML6075's operating voltage ranges from 1.7 V to 3.6 V.

Datasheet:
- [VEML6075](https://cdn.sparkfun.com/assets/3/c/3/2/f/veml6075.pdf)

Application note:
- [Designing the VEML6075 into an Application](https://cdn.sparkfun.com/assets/3/9/d/4/1/designingveml6075.pdf)

## Usage

To use this driver, import this crate and an `embedded_hal` implementation,
then instantiate the device.

Please find additional examples using hardware in this repository: [driver-examples]

[driver-examples]: https://github.com/eldruin/driver-examples

```rust
use linux_embedded_hal::I2cdev;
use veml6075::{Calibration, Veml6075};

fn main() {
    let dev = I2cdev::new("/dev/i2c-1").unwrap();
    let mut sensor = Veml6075::new(dev, Calibration::default());
    let m = sensor.read().unwrap();
    println!("UVA: {:2}, UVB: {:2}, UVI: {:2}", m.uva, m.uvb, m.uv_index);
}
```

## Support

For questions, issues, feature requests, and other changes, please file an
[issue in the github project](https://github.com/eldruin/veml6075-rs/issues).

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
   http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or
   http://opensource.org/licenses/MIT)

at your option.

### Contributing

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

...

## [0.2.1] - 2021-06-06

### Changed
- Simplified code in examples

## [0.2.0] - 2019-09-28

This driver is now functionally complete.

### Added
- Method to read calibrated measurement of UVA, UVB and UV index.
- Methods to set the integration time and dynamic setting.
- Method to change the operating mode and triggering a measurement when on active force mode.

### Changed
- [breaking-change] Renamed VEML6075 -> Veml6075 to comply with Rust naming conventions.
- [breaking-change] Renamed methods to read raw measurements

## [0.1.0] - 2018-10-13

This is the initial release to crates.io. All changes will be documented in this CHANGELOG.

[Unreleased]: https://github.com/eldruin/veml6075-rs/compare/v0.2.1...HEAD
[0.2.1]: https://github.com/eldruin/veml6075-rs/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/eldruin/veml6075-rs/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/eldruin/veml6075-rs/releases/tag/v0.1.0


tests/tests.rs
use embedded_hal_mock::i2c::{Mock as I2cMock, Transaction as I2cTrans};
use veml6075::{
    Calibration, DynamicSetting as DS, IntegrationTime as IT, Measurement, Mode, Veml6075,
};

const DEVICE_ADDRESS: u8 = 0x10;
struct Register;
impl Register {
    const CONFIG: u8 = 0x00;
    const UVA: u8 = 0x07;
    const UVB: u8 = 0x09;
    const UVCOMP1: u8 = 0x0A;
    const UVCOMP2: u8 = 0x0B;
    const DEVICE_ID: u8 = 0x0C;
}

pub fn new(transactions: &[I2cTrans]) -> Veml6075<I2cMock> {
    Veml6075::new(I2cMock::new(transactions), Calibration::default())
}

pub fn destroy(sensor: Veml6075<I2cMock>) {
    sensor.destroy().done();
}

macro_rules! cfg_test {
    ($name:ident, $method:ident, $value:expr $(, $arg:expr)* ) => {
        #[test]
        fn $name() {
            let transactions = [I2cTrans::write(
                DEVICE_ADDRESS,
                vec![Register::CONFIG, $value, 0],
            )];
            let mut dev = new(&transactions);
            dev.$method($($arg),*).unwrap();
            destroy(dev);
        }
    };
}

cfg_test!(can_enable, enable, 0);
cfg_test!(can_disable, disable, 1);
cfg_test!(set_it_50, set_integration_time, 1, IT::Ms50);
cfg_test!(set_it_100, set_integration_time, 0b0001_0001, IT::Ms100);
cfg_test!(set_it_200, set_integration_time, 0b0010_0001, IT::Ms200);
cfg_test!(set_it_400, set_integration_time, 0b0011_0001, IT::Ms400);
cfg_test!(set_it_800, set_integration_time, 0b0100_0001, IT::Ms800);
cfg_test!(set_ds_normal, set_dynamic_setting, 1, DS::Normal);
cfg_test!(set_ds_high, set_dynamic_setting, 0b0000_1001, DS::High);
cfg_test!(set_continuous, set_mode, 1, Mode::Continuous);
cfg_test!(set_active_force, set_mode, 0b0000_0011, Mode::ActiveForce);
cfg_test!(can_trigger, trigger_measurement, 0b0000_0101);

macro_rules! read_test {
    ($name:ident, $method:ident, $reg:ident) => {
        #[test]
        fn $name() {
            let transactions = [I2cTrans::write_read(
                DEVICE_ADDRESS,
                vec![Register::$reg],
                vec![0xCD, 0xAB],
            )];
            let mut dev = new(&transactions);
            let reading = dev.$method().unwrap();
            assert_eq!(reading, 0xABCD);
            destroy(dev);
        }
    };
}

read_test!(can_read_uva, read_uva_raw, UVA);
read_test!(can_read_uvb, read_uvb_raw, UVB);
read_test!(can_read_uvcomp1, read_uvcomp1_raw, UVCOMP1);
read_test!(can_read_uvcomp2, read_uvcomp2_raw, UVCOMP2);
read_test!(can_read_dev_id, read_device_id, DEVICE_ID);

#[test]
fn can_read_calibrated() {
    let transactions = [
        I2cTrans::write_read(DEVICE_ADDRESS, vec![Register::UVA], vec![0x7F, 0x0F]),
        I2cTrans::write_read(DEVICE_ADDRESS, vec![Register::UVB], vec![0xBA, 0x16]),
        I2cTrans::write_read(DEVICE_ADDRESS, vec![Register::UVCOMP1], vec![0xEF, 0x03]),
        I2cTrans::write_read(DEVICE_ADDRESS, vec![Register::UVCOMP2], vec![0xD7, 0x02]),
    ];
    let mut dev = new(&transactions);
    let Measurement { uva, uvb, uv_index } = dev.read().unwrap();

    let expected_uva = 3967.0 - 2.22 * 1007.0 - 1.33 * 727.0;
    assert!(uva - 0.5 < expected_uva);
    assert!(uva + 0.5 > expected_uva);
    let expected_uvb = 5818.0 - 2.95 * 1007.0 - 1.74 * 727.0;
    assert!(uvb - 0.5 < expected_uvb);
    assert!(uvb + 0.5 > expected_uvb);
    let expected_uv_index = (uva * 0.001_461 + uvb * 0.002_591) / 2.0;
    assert!(uv_index - 0.5 < expected_uv_index);
    assert!(uv_index + 0.5 > expected_uv_index);

    destroy(dev);
}

#[test]
fn calibration_default() {
    let c = Calibration {
        uva_visible: 2.22,
        uva_ir: 1.33,
        uvb_visible: 2.95,
        uvb_ir: 1.74,
        uva_responsivity: 0.001_461,
        uvb_responsivity: 0.002_591,
    };
    assert_eq!(c, Calibration::default());
}

#[test]
fn measurement_can_store() {
    let m = Measurement {
        uva: 1.1,
        uvb: 2.2,
        uv_index: 3.3,
    };
    assert!(m.uva - 0.5 < 1.1);
    assert!(m.uva + 0.5 > 1.1);
    assert!(m.uvb - 0.5 < 2.2);
    assert!(m.uvb + 0.5 > 2.2);
    assert!(m.uv_index - 0.5 < 3.3);
    assert!(m.uv_index + 0.5 > 3.3);
    assert_eq!(
        m,
        Measurement {
            uva: 1.1,
            uvb: 2.2,
            uv_index: 3.3,
        }
    );
}


src/lib.rs
//! This is a platform agnostic Rust driver for the VEML6075 UVA and UVB
//! light sensor, based on the [`embedded-hal`] traits.
//!
//! [`embedded-hal`]: https://github.com/rust-embedded/embedded-hal
//!
//! This driver allows you to:
//! - Enable/disable the sensor. See: [`enable()`].
//! - Read calibrated UVA, UVB and UV index measurement. See: [`read()`].
//! - Read raw measurement. See: [`read_uva_raw()`].
//! - Set integration time. See: [`set_integration_time()`].
//! - Set dynamic setting. See: [`set_dynamic_setting()`].
//! - Change operating mode. See: [`set_mode()`].
//! - Trigger measurement when on active force mode. See: [`trigger_measurement()`].
//! - Read the device id. See: [`read_device_id()`].
//!
//! [Introductory blog post](https://blog.eldruin.com/veml6075-uva-uvb-uv-index-light-sensor-driver-in-rust/)
//!
//! [`enable()`]: struct.Veml6075.html#method.enable
//! [`read()`]: struct.Veml6075.html#method.read
//! [`read_uva_raw()`]: struct.Veml6075.html#method.read_uva_raw
//! [`set_integration_time()`]: struct.Veml6075.html#method.set_integration_time
//! [`set_dynamic_setting()`]: struct.Veml6075.html#method.set_dynamic_setting
//! [`set_mode()`]: struct.Veml6075.html#method.set_mode
//! [`trigger_measurement()`]: struct.Veml6075.html#method.trigger_measurement
//! [`read_device_id()`]: struct.Veml6075.html#method.read_device_id
//!
//! ## The device
//! The VEML6075 senses UVA and UVB light and incorporates photodiode,
//! amplifiers,and analog / digital circuits into a single chip using a
//! CMOS process. When the UV sensor is applied, it is able to detect
//! UVA and UVB intensity to provide a measure of the signal strength
//! as well as allowing for UVI measurement.
//! The VEML6075 provides excellent temperature compensation capability
//! for keeping the output stable under changing temperature.
//! VEML6075's functionality is easily operated via the simple command
//! format of I2C (SMBus compatible) interface protocol.
//! VEML6075's operating voltage ranges from 1.7 V to 3.6 V.
//!
//! Datasheet:
//! - [VEML6075](https://cdn.sparkfun.com/assets/3/c/3/2/f/veml6075.pdf)
//!
//! Application note:
//! - [Designing the VEML6075 into an Application](https://cdn.sparkfun.com/assets/3/9/d/4/1/designingveml6075.pdf)
//!
//! ## Usage examples (see also examples folder)
//!
//! Please find additional examples using hardware in this repository: [driver-examples]
//!
//! [driver-examples]: https://github.com/eldruin/driver-examples
//!
//! ### Read calibrated UVA, UVB and UV index
//!
//! Import this crate and an `embedded_hal` implementation, then instantiate
//! the device:
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use veml6075::{Calibration, Veml6075};
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Veml6075::new(dev, Calibration::default());
//! let m = sensor.read().unwrap();
//! println!("UVA: {:2}, UVB: {:2}, UVI: {:2}", m.uva, m.uvb, m.uv_index);
//! ```
//!
//! ### Set integration time to 400ms
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use veml6075::{Calibration, IntegrationTime, Veml6075};
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Veml6075::new(dev, Calibration::default());
//! sensor.set_integration_time(IntegrationTime::Ms400).unwrap();
//! ```
//!
//! ### Set high dynamic setting
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use veml6075::{Calibration, DynamicSetting, Veml6075};
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Veml6075::new(dev, Calibration::default());
//! sensor.set_dynamic_setting(DynamicSetting::High).unwrap();
//! ```
//!
//! ### Change mode to active force (one-shot) and trigger a measurement
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use veml6075::{Calibration, Mode, Veml6075};
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Veml6075::new(dev, Calibration::default());
//! sensor.set_mode(Mode::ActiveForce).unwrap();
//! loop {
//!     sensor.trigger_measurement().unwrap();
//!     // wait until measurement is ready (integration time)
//!     let m = sensor.read().unwrap();
//!     println!("Measurements UVA: {:2}, UVB: {:2}", m.uva, m.uvb);
//! }
//! ```
//!
//! ### Read raw measurements for UVA and UVB
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use veml6075::{Calibration, Veml6075};
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Veml6075::new(dev, Calibration::default());
//! let uva = sensor.read_uva_raw().unwrap();
//! let uvb = sensor.read_uvb_raw().unwrap();
//! println!("Measurements UVA: {}, UVB: {}", uva, uvb);
//! ```

#![doc(html_root_url = "https://docs.rs/veml6075/0.2.1")]
#![deny(unsafe_code)]
#![deny(missing_docs)]
#![no_std]

/// All possible errors in this crate
#[derive(Debug)]
pub enum Error<E> {
    /// I²C bus error
    I2C(E),
}

/// Calibrated Measurement
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Measurement {
    /// UVA calibrated reading
    pub uva: f32,
    /// UVB calibrated reading
    pub uvb: f32,
    /// UV index
    pub uv_index: f32,
}

/// Integration time
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum IntegrationTime {
    /// 50 ms
    Ms50,
    /// 100 ms
    Ms100,
    /// 200 ms
    Ms200,
    /// 400 ms
    Ms400,
    /// 800 ms
    Ms800,
}

/// Dynamic setting
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum DynamicSetting {
    /// Normal dynamic setting
    Normal,
    /// High dynamic setting
    High,
}

/// Operating mode
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Mode {
    /// Continuous measurement (default)
    Continuous,
    /// Active force (one-shot)
    ActiveForce,
}

/// Calibration coefficients
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Calibration {
    /// UVA visible (a) coefficient
    pub uva_visible: f32,
    /// UVA IR (b) coefficient
    pub uva_ir: f32,
    /// UVB visible (c) coefficient
    pub uvb_visible: f32,
    /// UVB IR (d) coefficient
    pub uvb_ir: f32,
    /// UVA responsivity
    pub uva_responsivity: f32,
    /// UVB responsivity
    pub uvb_responsivity: f32,
}

/// Veml6075 device driver.
#[derive(Debug, Default)]
pub struct Veml6075<I2C> {
    /// The concrete I²C device implementation.
    i2c: I2C,
    /// Configuration register status.
    config: u8,
    calibration: Calibration,
}

mod device_impl;

impl Default for Calibration {
    fn default() -> Self {
        Calibration {
            uva_visible: 2.22,
            uva_ir: 1.33,
            uvb_visible: 2.95,
            uvb_ir: 1.74,
            uva_responsivity: 0.001_461,
            uvb_responsivity: 0.002_591,
        }
    }
}


src/device_impl.rs
//! Device implementation
use crate::{Calibration, DynamicSetting, Error, IntegrationTime, Measurement, Mode, Veml6075};
use embedded_hal::blocking::i2c::{Write, WriteRead};

struct Register;
impl Register {
    const CONFIG: u8 = 0x00;
    const UVA: u8 = 0x07;
    const UVB: u8 = 0x09;
    const UVCOMP1: u8 = 0x0A;
    const UVCOMP2: u8 = 0x0B;
    const DEVICE_ID: u8 = 0x0C;
}

struct BitFlags;
impl BitFlags {
    const SHUTDOWN: u8 = 0b0000_0001;
    const HD: u8 = 0b0000_1000;
    const UV_TRIG: u8 = 0b0000_0100;
    const UV_AF: u8 = 0b0000_0010;
}

const DEVICE_ADDRESS: u8 = 0x10;

impl<I2C, E> Veml6075<I2C>
where
    I2C: Write<Error = E>,
{
    /// Create new instance of the Veml6075 device.
    pub fn new(i2c: I2C, calibration: Calibration) -> Self {
        Veml6075 {
            i2c,
            config: 0x01, // shutdown
            calibration,
        }
    }

    /// Destroy driver instance, return I²C bus instance.
    pub fn destroy(self) -> I2C {
        self.i2c
    }

    /// Enable the sensor.
    pub fn enable(&mut self) -> Result<(), Error<E>> {
        let config = self.config;
        self.write_config(config & !BitFlags::SHUTDOWN)
    }

    /// Disable the sensor (shutdown).
    pub fn disable(&mut self) -> Result<(), Error<E>> {
        let config = self.config;
        self.write_config(config | BitFlags::SHUTDOWN)
    }

    /// Set operating mode
    pub fn set_mode(&mut self, mode: Mode) -> Result<(), Error<E>> {
        // This device does not report when a measurement is finished
        // so an non-blocking (`nb`) API is not possible to implement.
        // This is why I think the APIs are not different enough to
        // grant `into_one_shot()` and `into_continuous()` transformation
        // methods as the error handling for those is somewhat cumbersome.
        let config = match mode {
            Mode::Continuous => self.config & !BitFlags::UV_AF,
            Mode::ActiveForce => self.config | BitFlags::UV_AF,
        };
        self.write_config(config)
    }

    /// Trigger a measurement when on active force (one-shot) mode.
    pub fn trigger_measurement(&mut self) -> Result<(), Error<E>> {
        // this flag will automatically be set back to 0.
        let config = self.config | BitFlags::UV_TRIG;
        self.i2c
            .write(DEVICE_ADDRESS, &[Register::CONFIG, config, 0])
            .map_err(Error::I2C)
    }

    /// Set the integration time.
    pub fn set_integration_time(&mut self, it: IntegrationTime) -> Result<(), Error<E>> {
        let config = self.config & 0b1000_1111;
        let config = match it {
            IntegrationTime::Ms50 => config,
            IntegrationTime::Ms100 => config | 1 << 4,
            IntegrationTime::Ms200 => config | 2 << 4,
            IntegrationTime::Ms400 => config | 3 << 4,
            IntegrationTime::Ms800 => config | 4 << 4,
        };
        self.write_config(config)
    }

    /// Set the dynamic setting.
    pub fn set_dynamic_setting(&mut self, ds: DynamicSetting) -> Result<(), Error<E>> {
        let config = match ds {
            DynamicSetting::Normal => self.config & !BitFlags::HD,
            DynamicSetting::High => self.config | BitFlags::HD,
        };
        self.write_config(config)
    }

    fn write_config(&mut self, config: u8) -> Result<(), Error<E>> {
        self.i2c
            .write(DEVICE_ADDRESS, &[Register::CONFIG, config, 0])
            .map_err(Error::I2C)?;
        self.config = config;
        Ok(())
    }
}

impl<I2C, E> Veml6075<I2C>
where
    I2C: WriteRead<Error = E>,
{
    /// Read the sensor data and calculate calibrated reading values.
    pub fn read(&mut self) -> Result<Measurement, Error<E>> {
        let uva = self.read_uva_raw()?;
        let uvb = self.read_uvb_raw()?;
        let uvcomp1 = self.read_uvcomp1_raw()?;
        let uvcomp2 = self.read_uvcomp2_raw()?;
        let uva = f32::from(uva)
            - (self.calibration.uva_visible * f32::from(uvcomp1))
            - (self.calibration.uva_ir * f32::from(uvcomp2));
        let uvb = f32::from(uvb)
            - (self.calibration.uvb_visible * f32::from(uvcomp1))
            - (self.calibration.uvb_ir * f32::from(uvcomp2));
        let uv_index = (uva * self.calibration.uva_responsivity
            + uvb * self.calibration.uvb_responsivity)
            / 2.0;
        Ok(Measurement { uva, uvb, uv_index })
    }

    /// Read the raw UVA sensor data.
    pub fn read_uva_raw(&mut self) -> Result<u16, Error<E>> {
        self.read_register(Register::UVA)
    }

    /// Read the raw UVB sensor data.
    pub fn read_uvb_raw(&mut self) -> Result<u16, Error<E>> {
        self.read_register(Register::UVB)
    }

    /// Read the raw UVcomp1 sensor data.
    pub fn read_uvcomp1_raw(&mut self) -> Result<u16, Error<E>> {
        self.read_register(Register::UVCOMP1)
    }

    /// Read the raw UVcomp2 sensor data.
    pub fn read_uvcomp2_raw(&mut self) -> Result<u16, Error<E>> {
        self.read_register(Register::UVCOMP2)
    }

    /// Read the device ID
    pub fn read_device_id(&mut self) -> Result<u16, Error<E>> {
        self.read_register(Register::DEVICE_ID)
    }

    fn read_register(&mut self, register: u8) -> Result<u16, Error<E>> {
        let mut data = [0; 2];
        self.i2c
            .write_read(DEVICE_ADDRESS, &[register], &mut data)
            .map_err(Error::I2C)?;
        Ok(u16::from(data[1]) << 8 | u16::from(data[0]))
    }
}
